Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs

```diff
@@ -1564,6 +1524,59 @@ public override void Push()
       SendThisVC("(push 1)");
       DeclCollector.Push();
     }
+    
+    
+  
+  
+    public class ProverProblemCollector
```

Comment:
Could you move this class to a separate file? This file is already very large.

Reply:
Warnings are enqueued in the case of a parse error until they are displayed when an error occurs (in the case of an error). For example, in [`ParseRCount`]() within CheckSat
https://github.com/boogie-org/boogie/blob/7b389a524fe1cb475108688cb97b3a9ea1e7e639/Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs#L1544-L1558
warnings are processed in `HandleProverError`: 
https://github.com/boogie-org/boogie/blob/7b389a524fe1cb475108688cb97b3a9ea1e7e639/Source/Provers/SMTLib/SMTLibProcessTheoremProver.cs#L575

For errors, you're right, they are read nowhere except for their `.Count > 0` property.
So I made the field `Errors` private and added the public property `HadErrors` which reads nicely. I think it's not too costly to keep the errors in the concurrent stack anyway and it had been already useful for me in the past.

Reply:
Good insight, but `FlushProverWarnings()` is also called if there's no error right? It is called _before_ the check `if (error == "") return;`

Reply:
It seems that the `proverWarnings` are flushed using an `ErrorHandler handler` that is provided when `CheckSat` is called in either `SMTLibBatchTheoremProver` or `SMTLibInteractiveTheoremProver`

Removing `proverWarnings` by immediately using `currentErrorHandler` might result in nothing if `currentErrorHandler` is not set at best, and in a null pointer exception if we don't check this and it's null.

There is no unit test guarding us against failures of the underlying SMT solver. I implemented the change in the last commit, but I would like you to review it to ensure it is coherent.